File: src/engine/word-generator.ts
import { getLanguage } from "../data/languages/index.js";

export interface GeneratorConfig {
  language: string;
  count: number;
  punctuation: boolean;
  numbers: boolean;
}

export function generateWords(config: GeneratorConfig): string[] {
  const language = getLanguage(config.language);
  const { words: wordList } = language;
  const result: string[] = [];

  for (let i = 0; i < config.count; i++) {
    // Power-law distribution: favor frequent words (beginning of list)
    const index = Math.floor(Math.random() ** 1.5 * wordList.length);
    let word = wordList[index];

    // Numbers mode: ~8% chance to replace with a number
    if (config.numbers && Math.random() < 0.08) {
      const digits = Math.floor(Math.random() * 4) + 1;
      word = String(Math.floor(Math.random() * 10 ** digits));
      result.push(word);
      continue;
    }

    result.push(word);
  }

  if (config.punctuation) {
    return applyPunctuation(result);
  }

  return result;
}

function applyPunctuation(words: string[]): string[] {
  const result = [...words];
  let sentenceStart = true;

  for (let i = 0; i < result.length; i++) {
    // Capitalize first word of sentence
    if (sentenceStart && result[i].length > 0) {
      result[i] = result[i][0].toUpperCase() + result[i].slice(1);
      sentenceStart = false;
    }

    // End of sentence: period every 5-12 words
    if (i > 0 && Math.random() < 0.12) {
      result[i] = result[i] + ".";
      sentenceStart = true;
      continue;
    }

    // Comma: ~6% chance mid-sentence
    if (!sentenceStart && Math.random() < 0.06) {
      result[i] = result[i] + ",";
    }
  }

  return result;
}

File: src/engine/word_generator.rs
use crate::data::languages::get_language;
use crate::types::GeneratorConfig;
use rand::Rng;

pub fn generate_words(config: &GeneratorConfig) -> Vec<String> {
    let language = get_language(&config.language);
    let word_list = &language.words;
    let mut rng = rand::rng();
    let mut result = Vec::with_capacity(config.count);

    for _ in 0..config.count {
        // Power-law distribution: favor frequent words (beginning of list)
        let r: f64 = rng.random();
        let index = (r.powf(1.5) * word_list.len() as f64).floor() as usize;
        let index = index.min(word_list.len() - 1);
        let mut word = word_list[index].clone();

        // Numbers mode: ~8% chance to replace with a number
        if config.numbers && rng.random::<f64>() < 0.08 {
            let digits = rng.random_range(1..=4);
            let max = 10_u32.pow(digits);
            word = rng.random_range(0..max).to_string();
            result.push(word);
            continue;
        }

        result.push(word);
    }

    if config.punctuation {
        apply_punctuation(&mut result, &mut rng);
    }

    result
}

fn apply_punctuation(words: &mut [String], rng: &mut impl Rng) {
    let mut sentence_start = true;

    for i in 0..words.len() {
        // Capitalize first word of sentence
        if sentence_start && !words[i].is_empty() {
            let mut chars = words[i].chars();
            if let Some(first) = chars.next() {
                words[i] = first.to_uppercase().to_string() + chars.as_str();
            }
            sentence_start = false;
        }

        // End of sentence: period ~12% chance
        if i > 0 && rng.random::<f64>() < 0.12 {
            words[i].push('.');
            sentence_start = true;
            continue;
        }

        // Comma: ~6% chance mid-sentence
        if !sentence_start && rng.random::<f64>() < 0.06 {
            words[i].push(',');
        }
    }
}
